[Screen Time] Main Thread Checker: UI API called on a background thread: `-[WKWebView observeValueForKeyPath:ofObject:change:context:]`

pxlcoder opened the pull request:
#### 79c5ad58ede504e1d7e792d00e931a2acb99522a
<pre>
[Screen Time] Main Thread Checker: UI API called on a background thread: `-[WKWebView observeValueForKeyPath:ofObject:change:context:]`
<a href="https://bugs.webkit.org/show_bug.cgi?id=300904">https://bugs.webkit.org/show_bug.cgi?id=300904</a>
<a href="https://rdar.apple.com/162653113">rdar://162653113</a>

Reviewed by Abrar Rahman Protyasha.

300778@main added `WKWebView` as an observer of `STScreenTimeConfigurationObserver`,
which performs KVO on a background thread. This triggers the Main Thread Checker
as `WKWebView` is annotated for use on the main thread only. In reality, the
error is benign, since 300778@main ensured the actual logic is run on the main thread.

To avoid developer confusion due to this output, refactor the logic to occur
in a separate class, `WKScreenTimeConfigurationObserver`.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageControllerIfNeeded]):
(-[WKWebView _uninstallScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WKScreenTimeConfigurationObserver.h: Added.
* Source/WebKit/UIProcess/Cocoa/WKScreenTimeConfigurationObserver.mm: Added.
(-[WKScreenTimeConfigurationObserver initWithView:]):
(-[WKScreenTimeConfigurationObserver observeValueForKeyPath:ofObject:change:context:]):
(-[WKScreenTimeConfigurationObserver startObserving]):
(-[WKScreenTimeConfigurationObserver stopObserving]):
(-[WKScreenTimeConfigurationObserver enforcesChildRestrictions]):
* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:

Unified sources build fix.

* Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:

Unified sources build fix.

* Source/WebKit/UIProcess/mac/WebDateTimePickerMac.mm:

Unified sources build fix.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/301728@main">https://commits.webkit.org/301728@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b2fa477d75a9aca4ef53a35f6cc9a2982dcbd185

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126633 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37239 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133603 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78294 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/648c001b-bdb7-4588-8b2a-8e5941d6daf6) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46910 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54815 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96370 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64439 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/18084973-eafd-4eb3-9564-efbe2308cdaf) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129581 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76895 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9b1bf98d-a62d-452f-bc6f-212135f2141a) 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76999 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31734 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136173 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53321 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41022 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104884 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53809 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109611 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104604 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50065 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50778 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19844 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53243 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59056 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52519 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55856 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54257 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/Cocoa/WKScreenTimeConfigurationObserver.mm:
```diff
@@ -0,0 +1,88 @@
+
+- (void)startObserving
+{
+    _screenTimeConfigurationObserver = adoptNS([PAL::allocSTScreenTimeConfigurationObserverInstance() initWithUpdateQueue:globalDispatchQueueSingleton(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)]);
```
I know we only create `WKScreenTimeConfigurationObserver` if one doesn't exist, but maybe same idea with `STScreenTimeConfigurationObserver`? In case the caller does something redundant.

```objc
    if (!_screenTimeConfigurationObserver) {
        _screenTimeConfigurationObserver = ...
    }
    ...
```

pxlcoder replied:
Added an early return if it exists.

pxlcoder committed:
[Screen Time] Main Thread Checker: UI API called on a background thread: `-[WKWebView observeValueForKeyPath:ofObject:change:context:]`
https://bugs.webkit.org/show_bug.cgi?id=300904
rdar://162653113

Reviewed by Abrar Rahman Protyasha.

300778@main added `WKWebView` as an observer of `STScreenTimeConfigurationObserver`,
which performs KVO on a background thread. This triggers the Main Thread Checker
as `WKWebView` is annotated for use on the main thread only. In reality, the
error is benign, since 300778@main ensured the actual logic is run on the main thread.

To avoid developer confusion due to this output, refactor the logic to occur
in a separate class, `WKScreenTimeConfigurationObserver`.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageControllerIfNeeded]):
(-[WKWebView _uninstallScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WKScreenTimeConfigurationObserver.h: Added.
* Source/WebKit/UIProcess/Cocoa/WKScreenTimeConfigurationObserver.mm: Added.
(-[WKScreenTimeConfigurationObserver initWithView:]):
(-[WKScreenTimeConfigurationObserver observeValueForKeyPath:ofObject:change:context:]):
(-[WKScreenTimeConfigurationObserver startObserving]):
(-[WKScreenTimeConfigurationObserver stopObserving]):
(-[WKScreenTimeConfigurationObserver enforcesChildRestrictions]):
* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:

Unified sources build fix.

* Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:

Unified sources build fix.

* Source/WebKit/UIProcess/mac/WebDateTimePickerMac.mm:

Unified sources build fix.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/301728@main
